<a href="https://colab.research.google.com/github/Imadittyaa/coffee-sales/blob/main/coffee_sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
d1=pd.read_csv(r'https://github.com/Imadittyaa/coffee-sales/blob/main/index_1.csv')
d2=pd.read_csv(r'index_2.csv')
data=pd.concat([d1,d2],ignore_index=True)
data.shape

(3899, 6)

# Infomation of data

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3899 entries, 0 to 3898
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         3899 non-null   object 
 1   datetime     3898 non-null   object 
 2   cash_type    3898 non-null   object 
 3   card         3547 non-null   object 
 4   money        3898 non-null   float64
 5   coffee_name  3896 non-null   object 
dtypes: float64(1), object(5)
memory usage: 182.9+ KB


In [ ]:
data.head(n=21)

date        datetime cash_type                 card  money  \
0   3/1/2024  3/1/2024 10:15      card  ANON-0000-0000-0001   38.7   
1   3/1/2024  3/1/2024 12:19      card  ANON-0000-0000-0002   38.7   
2   3/1/2024  3/1/2024 12:20      card  ANON-0000-0000-0002   38.7   
3   3/1/2024  3/1/2024 13:46      card  ANON-0000-0000-0003   28.9   
4   3/1/2024  3/1/2024 13:48      card  ANON-0000-0000-0004   38.7   
5   3/1/2024  3/1/2024 15:39      card  ANON-0000-0000-0005   33.8   
6   3/1/2024  3/1/2024 16:19      card  ANON-0000-0000-0006   38.7   
7   3/1/2024  3/1/2024 18:39      card  ANON-0000-0000-0007   33.8   
8   3/1/2024  3/1/2024 19:22      card  ANON-0000-0000-0008   38.7   
9   3/1/2024  3/1/2024 19:23      card  ANON-0000-0000-0008   33.8   
10  3/1/2024  3/1/2024 19:29      card  ANON-0000-0000-0009   33.8   
11  3/2/2024  3/2/2024 10:22      card  ANON-0000-0000-0010   28.9   
12  3/2/2024  3/2/2024 10:30      cash                  NaN   40.0   
13  3/2/2024  3/2/2024 10:41      card  ANON-0000-0000-0011   33.8   
14  3/2/2024  3/2/2024 11:59      card  ANON-0000-0000-0012   33.8   
15  3/2/2024  3/2/2024 14:38      card  ANON-0000-0000-0013   28.9   
16  3/2/2024  3/2/2024 16:37      card  ANON-0000-0000-0014   33.8   
17  3/2/2024  3/2/2024 17:34      card  ANON-0000-0000-0004   28.9   
18  3/3/2024  3/3/2024 10:10      cash                  NaN   40.0   
19  3/3/2024  3/3/2024 10:27      card  ANON-0000-0000-0015   38.7   
20  3/3/2024  3/3/2024 11:33      card  ANON-0000-0000-0016   28.9   

            coffee_name  
0                 Latte  
1         Hot Chocolate  
2         Hot Chocolate  
3             Americano  
4                 Latte  
5   Americano with Milk  
6         Hot Chocolate  
7   Americano with Milk  
8                 Cocoa  
9   Americano with Milk  
10  Americano with Milk  
11            Americano  
12                Latte  
13  Americano with Milk  
14  Americano with Milk  
15            Americano  
16  Americano with Milk  
17            Americano  
18                Latte  
19                Latte  
20              Cortado

# Some operations on data

In [ ]:
data.rename(columns={'datetime':'time'},inplace=True)
data['time']=data['time'].str[-5:]

In [ ]:
data['date']=pd.to_datetime(data['date'],errors='coerce')

In [ ]:
cash_data=data[data['cash_type'].str.lower()=='cash'].reset_index(drop=True)
card_data=data[data['cash_type'].str.lower()=='card'].reset_index(drop=True)

**Cleared coffee sales data**

In [ ]:
data.head(n=21)

date   time cash_type                 card  money  \
0  2024-03-01  10:15      card  ANON-0000-0000-0001   38.7   
1  2024-03-01  12:19      card  ANON-0000-0000-0002   38.7   
2  2024-03-01  12:20      card  ANON-0000-0000-0002   38.7   
3  2024-03-01  13:46      card  ANON-0000-0000-0003   28.9   
4  2024-03-01  13:48      card  ANON-0000-0000-0004   38.7   
5  2024-03-01  15:39      card  ANON-0000-0000-0005   33.8   
6  2024-03-01  16:19      card  ANON-0000-0000-0006   38.7   
7  2024-03-01  18:39      card  ANON-0000-0000-0007   33.8   
8  2024-03-01  19:22      card  ANON-0000-0000-0008   38.7   
9  2024-03-01  19:23      card  ANON-0000-0000-0008   33.8   
10 2024-03-01  19:29      card  ANON-0000-0000-0009   33.8   
11 2024-03-02  10:22      card  ANON-0000-0000-0010   28.9   
12 2024-03-02  10:30      cash                  NaN   40.0   
13 2024-03-02  10:41      card  ANON-0000-0000-0011   33.8   
14 2024-03-02  11:59      card  ANON-0000-0000-0012   33.8   
15 2024-03-02  14:38      card  ANON-0000-0000-0013   28.9   
16 2024-03-02  16:37      card  ANON-0000-0000-0014   33.8   
17 2024-03-02  17:34      card  ANON-0000-0000-0004   28.9   
18 2024-03-03  10:10      cash                  NaN   40.0   
19 2024-03-03  10:27      card  ANON-0000-0000-0015   38.7   
20 2024-03-03  11:33      card  ANON-0000-0000-0016   28.9   

            coffee_name  
0                 Latte  
1         Hot Chocolate  
2         Hot Chocolate  
3             Americano  
4                 Latte  
5   Americano with Milk  
6         Hot Chocolate  
7   Americano with Milk  
8                 Cocoa  
9   Americano with Milk  
10  Americano with Milk  
11            Americano  
12                Latte  
13  Americano with Milk  
14  Americano with Milk  
15            Americano  
16  Americano with Milk  
17            Americano  
18                Latte  
19                Latte  
20              Cortado

# Data divide on basis of payment method

**cash_data actually denote to data on basis of payment by cash and card_data actually data on basis of payment by card**

In [ ]:
cash_data.head(n=21)

date   time cash_type card  money          coffee_name
0  2024-03-02  10:30      cash  NaN   40.0                Latte
1  2024-03-03  10:10      cash  NaN   40.0                Latte
2  2024-03-06  12:30      cash  NaN   35.0  Americano with Milk
3  2024-03-07  10:08      cash  NaN   40.0                Latte
4  2024-03-07  11:25      cash  NaN   40.0                Latte
5  2024-03-09  11:49      cash  NaN   40.0        Hot Chocolate
6  2024-03-10   7:44      cash  NaN   30.0            Americano
7  2024-03-10   7:45      cash  NaN   35.0  Americano with Milk
8  2024-03-10  10:06      cash  NaN   40.0                Latte
9  2024-03-10  17:51      cash  NaN   35.0  Americano with Milk
10 2024-03-11  10:18      cash  NaN   40.0                Latte
11 2024-03-11  11:24      cash  NaN   40.0           Cappuccino
12 2024-03-11  11:24      cash  NaN   30.0              Cortado
13 2024-03-11  11:26      cash  NaN   30.0            Americano
14 2024-03-12  10:15      cash  NaN   40.0                Latte
15 2024-03-13  12:34      cash  NaN   25.0             Espresso
16 2024-03-14  10:02      cash  NaN   40.0                Latte
17 2024-03-15  10:02      cash  NaN   40.0                Latte
18 2024-03-17  10:10      cash  NaN   40.0                Latte
19 2024-03-19  10:20      cash  NaN   30.0              Cortado
20 2024-03-19  19:26      cash  NaN   40.0        Hot Chocolate

In [ ]:
len(cash_data)

169

**sales by cash payment method is 169**

In [ ]:
card_data.head(n=21)

date   time cash_type                 card  money  \
0  2024-03-01  10:15      card  ANON-0000-0000-0001   38.7   
1  2024-03-01  12:19      card  ANON-0000-0000-0002   38.7   
2  2024-03-01  12:20      card  ANON-0000-0000-0002   38.7   
3  2024-03-01  13:46      card  ANON-0000-0000-0003   28.9   
4  2024-03-01  13:48      card  ANON-0000-0000-0004   38.7   
5  2024-03-01  15:39      card  ANON-0000-0000-0005   33.8   
6  2024-03-01  16:19      card  ANON-0000-0000-0006   38.7   
7  2024-03-01  18:39      card  ANON-0000-0000-0007   33.8   
8  2024-03-01  19:22      card  ANON-0000-0000-0008   38.7   
9  2024-03-01  19:23      card  ANON-0000-0000-0008   33.8   
10 2024-03-01  19:29      card  ANON-0000-0000-0009   33.8   
11 2024-03-02  10:22      card  ANON-0000-0000-0010   28.9   
12 2024-03-02  10:41      card  ANON-0000-0000-0011   33.8   
13 2024-03-02  11:59      card  ANON-0000-0000-0012   33.8   
14 2024-03-02  14:38      card  ANON-0000-0000-0013   28.9   
15 2024-03-02  16:37      card  ANON-0000-0000-0014   33.8   
16 2024-03-02  17:34      card  ANON-0000-0000-0004   28.9   
17 2024-03-03  10:27      card  ANON-0000-0000-0015   38.7   
18 2024-03-03  11:33      card  ANON-0000-0000-0016   28.9   
19 2024-03-03  12:26      card  ANON-0000-0000-0012   28.9   
20 2024-03-03  13:09      card  ANON-0000-0000-0017   38.7   

            coffee_name  
0                 Latte  
1         Hot Chocolate  
2         Hot Chocolate  
3             Americano  
4                 Latte  
5   Americano with Milk  
6         Hot Chocolate  
7   Americano with Milk  
8                 Cocoa  
9   Americano with Milk  
10  Americano with Milk  
11            Americano  
12  Americano with Milk  
13  Americano with Milk  
14            Americano  
15  Americano with Milk  
16            Americano  
17                Latte  
18              Cortado  
19            Americano  
20        Hot Chocolate

In [ ]:
len(card_data)

3729

**Sales by card payment method is 3729**

In [ ]:
plt.figure(figsize=(5,5))
plt.pie(
    [len(cash_data),len(card_data)],
    labels=['Cash','Card'],
    autopct='%1.1f%%',
    startangle=90,
    colors=['red','white'])
plt.title('Sales By Payment Method',fontsize=14)
plt.tight_layout()
plt.show()

# on comparing of cash_data and card_data

In [ ]:
cash_prices=cash_data.groupby('coffee_name')['money'].apply(lambda x:sorted (set(x))).reset_index(name='cash_prices')

In [ ]:
card_prices=card_data.groupby('coffee_name')['money'].apply(lambda x:sorted (set(x))).reset_index(name='card_prices')

In [ ]:
new_data=pd.merge(cash_prices,card_prices,on='coffee_name',how='outer')
new_data

coffee_name         cash_prices  \
0                    Americano  [25.0, 29.0, 30.0]   
1          Americano with Milk        [34.0, 35.0]   
2          Americano with milk              [25.0]   
3                   Cappuccino  [30.0, 39.0, 40.0]   
4                      Caramel              [22.0]   
5               Caramel coffee                 NaN   
6   Caramel with Irish whiskey                 NaN   
7       Caramel with chocolate                 NaN   
8            Caramel with milk              [25.0]   
9                    Chocolate                 NaN   
10       Chocolate with coffee              [30.0]   
11         Chocolate with milk              [27.0]   
12                       Cocoa        [39.0, 40.0]   
13   Coffee with Irish whiskey              [29.0]   
14       Coffee with chocolate              [28.0]   
15                     Cortado              [30.0]   
16        Double Irish whiskey              [28.0]   
17            Double chocolate              [25.0]   
18             Double espresso              [27.0]   
19   Double espresso with milk                 NaN   
20            Double ristretto                 NaN   
21              Double vanilla              [20.0]   
22                    Espresso  [20.0, 24.0, 25.0]   
23               Hot Chocolate        [39.0, 40.0]   
24               Hot milkshake                 NaN   
25               Irish whiskey              [25.0]   
26     Irish whiskey with milk              [27.0]   
27        Irish with chocolate                 NaN   
28                       Latte  [33.0, 39.0, 40.0]   
29                  Mochaccino              [30.0]   
30             Super chocolate              [28.0]   
31                         Tea              [15.0]   
32              Vanilla coffee                 NaN   
33  Vanilla with Irish whiskey              [28.0]   

                                card_prices  
0         [23.02, 25.0, 25.96, 27.92, 28.9]  
1               [27.92, 30.86, 32.82, 33.8]  
2                                    [25.0]  
3         [30.0, 32.82, 35.76, 37.72, 38.7]  
4                                       NaN  
5                                    [28.0]  
6                                    [28.0]  
7                                    [28.0]  
8                                    [25.0]  
9                                    [25.0]  
10                                   [30.0]  
11                                   [27.0]  
12              [32.82, 35.76, 37.72, 38.7]  
13                                   [29.0]  
14                                      NaN  
15              [23.02, 25.96, 27.92, 28.9]  
16                                   [28.0]  
17                                      NaN  
18                                      NaN  
19                                   [32.0]  
20                                   [27.0]  
21                                   [27.0]  
22  [18.12, 20.0, 21.06, 23.02, 24.0, 33.0]  
23              [32.82, 35.76, 37.72, 38.7]  
24                                   [30.0]  
25                                   [25.0]  
26                                   [27.0]  
27                                   [27.0]  
28        [32.82, 33.0, 35.76, 37.72, 38.7]  
29                                   [30.0]  
30                                   [28.0]  
31                                   [15.0]  
32                                   [27.0]  
33                                   [28.0]

In [ ]:
data.groupby('coffee_name')['money'].mean()

coffee_name
Americano                     26.032479
Americano with Milk           30.666408
Americano with milk           25.000000
Cappuccino                    35.804535
Caramel                       22.000000
Caramel coffee                28.000000
Caramel with Irish whiskey    28.000000
Caramel with chocolate        28.000000
Caramel with milk             25.000000
Chocolate                     25.000000
Chocolate with coffee         30.000000
Chocolate with milk           27.000000
Cocoa                         35.712593
Coffee with Irish whiskey     29.000000
Coffee with chocolate         28.000000
Cortado                       25.804315
Double Irish whiskey          28.000000
Double chocolate              25.000000
Double espresso               27.000000
Double espresso with milk     32.000000
Double ristretto              27.000000
Double vanilla                23.500000
Espresso                      20.968947
Hot Chocolate                 36.072553
Hot milkshake               

In [ ]:
plt.figure(figsize=(5,5))
plt.bar(x=new_data['coffee_name'],data.groupby('coffee_name')['money'].mea
plt.xlabel('COFFEE NAME')
plt.ylabel('PRICES')
plt.title('REPRESENTATION OF CASH_DATA AND CARD_DATA')
plt.show()

TypeError: can only concatenate list (not "int") to list

**There are several prices of each coffee-type**

# Total revenue and average transaction

In [ ]:
data['money'].agg(['sum','mean'])

sum     122321.5800
mean        31.3806
Name: money, dtype: float64

**Total revenue = 122321.58 and average transaction per coffee = 31.38**

# Favorite coffee

In [ ]:
data['coffee_name'].value_counts().head(n=10)

coffee_name
Americano with Milk    824
Latte                  806
Americano              593
Cappuccino             516
Cortado                292
Hot Chocolate          282
Cocoa                  243
Espresso               152
Americano with milk     44
Irish whiskey           20
Name: count, dtype: int64

# Revenue by date

In [ ]:
daily_revenue=data.groupby('date',sort=False).sum('money')
daily_revenue

money
date              
2024-03-01  396.30
2024-03-02  228.10
2024-03-03  349.10
2024-03-04  135.20
2024-03-05  338.50
...            ...
2025-03-19  763.56
2025-03-20  685.60
2025-03-21  741.80
2025-03-22  613.42
2025-03-23  419.76

[381 rows x 1 columns]

In [ ]:
daily_revenue.mean()

money    321.054016
dtype: float64

**Average daily transaction = 321.05**

In [ ]:
pd.to_datetime(data['time'],format='%H:%M',errors='coerce').dt.hour.value_counts()

time
10.0    363
11.0    316
16.0    306
12.0    271
17.0    268
15.0    267
19.0    253
18.0    247
14.0    245
13.0    240
21.0    210
20.0    177
22.0    123
23.0      3
Name: count, dtype: int64

**peak selling timing is 10 to 11 with 363**

# Save cleaned data as csv

In [ ]:
data.to_csv('coffee_data.csv',index=False)

# Average sale amount per coffee type

In [ ]:
data.groupby('coffee_name',sort=False).mean('money').round(decimals=2)

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


money
coffee_name                      
Latte                       35.56
Hot Chocolate               36.07
Americano                   26.03
Americano with Milk         30.67
Cocoa                       35.71
Cortado                     25.80
Espresso                    20.97
Cappuccino                  35.80
Tea                         15.00
Chocolate with coffee       30.00
Chocolate with milk         27.00
Coffee with Irish whiskey   29.00
Irish whiskey with milk     27.00
Double Irish whiskey        28.00
Double espresso             27.00
Americano with milk         25.00
Double espresso with milk   32.00
Caramel with Irish whiskey  28.00
Super chocolate             28.00
Double vanilla              23.50
Mochaccino                  30.00
Caramel with chocolate      28.00
Irish whiskey               25.00
Caramel coffee              28.00
Vanilla with Irish whiskey  28.00
Chocolate                   25.00
Caramel with milk           25.00
Coffee with chocolate       28.00
Irish with chocolate        27.00
Vanilla coffee              27.00
Double ristretto            27.00
Hot milkshake               30.00
Double chocolate            25.00
Caramel                     22.00

# Date with maximum sales

In [ ]:
daily_revenue.max()

money    979.08
dtype: float64

In [ ]:
daily_revenue[daily_revenue.values==daily_revenue.max().values].index

DatetimeIndex(['2025-02-21'], dtype='datetime64[ns]', name='date', freq=None)

**At 21/02/2025, Maximum sales(max revenue = 979.08 )**

# Peak selling day

In [ ]:
data['day']=data['date'].dt.day_name()
data['day'].value_counts()

day
Tuesday      604
Monday       591
Friday       586
Wednesday    553
Thursday     540
Saturday     531
Sunday       493
Name: count, dtype: int64

**Peak selling day is Tuesday with 604 sales**

In [ ]:
plt.figure(figsize=(5,5))
plt.plot(
